File: adxl345_python/adxl345.py
#!/usr/bin/env python

import smbus
import time
import sys

bus = smbus.SMBus(1)
address = 0x53

DR_200HZ   = 0x0B

DEVID       = 0x00
OFSX        = 0x1E
OFSY        = 0x1F
OFSZ        = 0x20
DUR         = 0x21
LATENT      = 0x22
WINDOW      = 0x23
INT_ENABLE  = 0x2E
BW_RATE     = 0x2C
POWER_CTL   = 0x2D
DATA_FORMAT = 0x31
DATAX0      = 0x32
DATAX1      = 0x33
DATAY0      = 0x34
DATAY1      = 0x35
DATAZ0      = 0x36
DATAZ1      = 0x37

def readDevId():
    return bus.read_byte_data(address, 0x00)

def setup():
    id = readDevId()
    if id != 0xE5:
        print('Error: device not detected!')
        return False
    # Put in stand by mode
    bus.write_byte_data(address, POWER_CTL, 0x00)
    # OFS, DUR, Latent, Window = 0 as we don't need it
    bus.write_i2c_block_data(address, OFSX, [0x00] * 6)
    # Set power mode and data rate
    bus.write_byte_data(address, BW_RATE, DR_200HZ)
    # Disable interrupts
    bus.write_byte_data(address, INT_ENABLE, 0x00)
    # Set our data format (+- 4g)
    bus.write_byte_data(address, DATA_FORMAT, 0x09)
    # Go into measurement mode
    bus.write_byte_data(address, POWER_CTL, 0x08)
    return True

def shutdown():
    # Go into stand by mode
    bus.write_byte_data(address, POWER_CTL, 0x00)
    return True

# 4mg per LSB
G_UNIT = (0.004)

print('Hello, World!')

setup()
print('Setup complete')

try:
    while True:
        data = bus.read_i2c_block_data(address, DATAX0, 6)
        x = (data[0] << 8) + data[1]
        y = (data[2] << 8) + data[3]
        z = (data[4] << 8) + data[5]
        if x >= 1<<15: x -= 1<<16
        if y >= 1<<15: y -= 1<<16
        if z >= 1<<15: z -= 1<<16
        x *= G_UNIT
        y *= G_UNIT
        z *= G_UNIT
        sys.stdout.write('\r{0:10.4f} {1:10.4f} {2:10.4f}'.format(x, y, z))
        sys.stdout.flush()

except KeyboardInterrupt:
    print('\nInterrupt signal received')

shutdown()
print('Shutdown complete')

File: adxl345_c/adxl345.c
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define ADDRESS 0x53

#define DR           0x0B
#define DEVID        0x00
#define OFSX         0x1E
#define OFSY         0x1F
#define OFSZ         0x20
#define DUR          0x21
#define LATENT       0x22
#define WINDOW       0x23
#define INT_ENABLE   0x2E
#define BW_RATE      0x2C
#define POWER_CTL    0x2D
#define DATA_FORMAT  0x31
#define DATAXA       0x32
#define DATAXB       0x33
#define DATAYA       0x34
#define DATAYB       0x35
#define DATAZA       0x36
#define DATAZB       0x37

#define G_UNIT       0.004

static const char *devName = "/dev/i2c-1";

void write_reg(int fd, char reg, char val) {
    char ar[2];
    ar[0] = reg;
    ar[1] = val;
    write(fd, ar, 2);
}

int main(int argc, char** argv) {
    int file;
    printf("I2C: connecting\n");
    if((file = open(devName, O_RDWR)) < 0) {
        fprintf(stderr, "I2C: failed to access %s\n", devName);
        exit(1);
    }
    printf("I2C: acquiring bus to 0x%x\n", ADDRESS);
    if(ioctl(file, I2C_SLAVE, ADDRESS) < 0) {
        fprintf(stderr, "I2C: failed to acquire bus\n");
        exit(1);
    }

    write_reg(file, POWER_CTL, 0x00);
    write_reg(file, OFSX, 0x00);
    write_reg(file, OFSY, 0x00);
    write_reg(file, OFSZ, 0x00);
    write_reg(file, DUR, 0x00);
    write_reg(file, LATENT, 0x00);
    write_reg(file, WINDOW, 0x00);
    write_reg(file, BW_RATE, DR);
    write_reg(file, INT_ENABLE, 0x00);
    write_reg(file, DATA_FORMAT, 0x09);
    write_reg(file, POWER_CTL, 0x08);

    unsigned char buf[6] = { 0x00 };
    
    while(1) {
        char addr = DATAXA;
        if(write(file, &addr, 1) != 1) {
            fprintf(stderr, "I2C: error writing reg address\n");
            continue;
        }
        if(read(file, buf, 6) != 6) {
            fprintf(stderr, "I2C: error reading data\n");
            continue;
        }
        float x = (buf[1] << 8) + buf[0];
        float y = (buf[3] << 8) + buf[2];
        float z = (buf[5] << 8) + buf[4];
        if(x >= 1<<15) x -= 1<<16;
        if(y >= 1<<15) y -= 1<<16;
        if(z >= 1<<15) z -= 1<<16;
        x *= G_UNIT;
        y *= G_UNIT;
        z *= G_UNIT;
        printf("\r%+7.5f %+7.5f %+7.5f", x, y, z);
        usleep(5);
    }
    
    return 0;
}
